nemo_aligner/data/nlp/builders.py

```diff
@@ -341,7 +341,7 @@ def build_dataloader(
     pad_samples_to_global_batch_size=False,
     collate_fn=None,
     load_gbs=True,
-    use_random_sampler=True,
+    use_random_sampler=False,
```

Comment:
Is there a reason to set default as False?

Reply:
Yes, if this is not set to False, then it will jumble all of the samples up and two responses to the same prompt will not appear in the same microbatch, which is needed for xxx_BT

Reply:
@gshennvm I think this will also affect dpo, ppo etc, is it ok to set as False?

Reply:
this is tricky... for the other algorithms it is technically ok to set it to False because we do "pairs" of dataset rather than single samples. but this changes the default behavior and means users cannot run multiple epochs without changing the code.


how about we push a flag into the yaml that defaults to True, but we can set this to False? something like model.data.train_ds.use_random_sampler

Reply:
Why does this affect multiple epoch training? 

In general, I think perhaps the easier approach is to simply have it as default shuffle = False across all algorithms. If user wants to shuf the training set, they can do it by themselves (e.g. with shuf on bash beforehand during data preparation). This reduces replication issues due to differences in shuffling on different settings (this way for shuffling was only introduced in 24.05 release AND in the future, we cannot guarantee that it will be done in the same way).

Also, right now, there is no way to disable shuffling across all algorithms to the best of my knowledge.

Reply:
yeah right now it's not great. we want to enable random shuffle because if the user naively trains for many epochs without random shuffle then the next epoch will repeat in the same order if we don't use this random shuffle

Reply:
```if the user naively trains for many epochs without random shuffle then the next epoch will repeat in the same order ``` --> this is actually the expected behavior [if each sample is represented by a letter i.e. DACB DACB DACB  is the right order rather than DACB CADB BACD ] The rationale is that If I trained for 1.5 epochs, and the job ended and then I resume training, I want to be sure that the next 0.5 epochs is something that was not seen yet in the second epoch) . Why is it bad to sample in the same order?

Reply:
` this is actually the expected behavior` -> I don't think people have aligned on this. Random order for each epoch may help alleviate overfitting in some cases. (e.g. I heard the long context team wants this behaviour)

Reply:
After discussions with @odelalleau and @gshennvm , we will not change the default behavior but make the shuffling behavior of both train and val user configurable.